File: python/examples/tools/parallel.py
import math

from mirascope import llm


@llm.tool
def sqrt_tool(number: float) -> float:
    """Computes the square root of a number"""
    return math.sqrt(number)


@llm.call("openai/gpt-5-mini", tools=[sqrt_tool])
def math_assistant(query: str):
    return query


response = math_assistant("What are the square roots of 3737, 4242, and 6464?")

tool_outputs = response.execute_tools()  # Execute all of the response's tool calls

answer = response.resume(tool_outputs)
print(answer.pretty())

# The square roots (approximate) are:
# - sqrt(3737) ≈ 61.1310068623117
# - sqrt(4242) ≈ 65.13063795173512
# - sqrt(6464) ≈ 80.39900496896712

File: typescript/examples/tools/parallel.ts
/**
 * Parallel tool execution example.
 *
 * Demonstrates executing multiple tool calls in parallel when the model
 * requests multiple tools at once.
 *
 * Run with: bun run example examples/tools/parallel.ts
 */
import { llm } from "mirascope";
import { z } from "zod";

const sqrtTool = llm.defineTool({
  name: "sqrt_tool",
  description: "Computes the square root of a number",
  validator: z.object({
    number: z.number().describe("The number to compute the square root of"),
  }),
  tool: ({ number }) => Math.sqrt(number),
});

const mathAssistant = llm.defineCall<{ query: string }>({
  model: "openai/gpt-4o-mini",
  tools: [sqrtTool],
  template: ({ query }) => query,
});

const response = await mathAssistant({
  query: "What are the square roots of 3737, 4242, and 6464?",
});

// executeTools() handles all tool calls in parallel
const toolOutputs = await response.executeTools();

const answer = await response.resume(toolOutputs);
console.log(answer.text());
// The square roots (approximate) are:
// - sqrt(3737) ≈ 61.13
// - sqrt(4242) ≈ 65.13
// - sqrt(6464) ≈ 80.40
